### ISSUE TITLE
add certificate as non Administrator

### ISSUE THREAD

**Issue by schnedann** [msys2/MSYS2-packages#1781]
the sittuation is simple.
I use msys2 at work but now my company changed things and one needs a certificate to access the internet.
updates from msys via pacman now do not work, as the certificate is missing in msys.
installing the certificate fails as only root can do this and root in msys2 terms seems to be the Administrator account.
I am a local admin on my machine with my normal user name, but the Adminstrator does not exist nor is it possible to have someone creating it - this our IT department will not do. There is only the ActiveDirectory, the network administrators and the local administrators. And beeing a local admin causes the commands to install certificates in msys to fail.

So far a posibility is needed to add certificates even as a user which is NOT the Administrator account and second it must be possible to update an existing installation without the need of pacman accessing the net or at least (re-)install something whithout the need of the certificates.

to summarize: currently I can not update any more and I guess also a reinstall will gracefully fail without having the needed certificates installed early in the process.
And equalize the Administrator with root ist not a viable option in AD managed scenarios

**Comment by Alexpux** [comment]
What problem install MSYS2 as regular user, not as administrator?

**Comment by schnedann** [comment]
Again: Msys2 is already installed. But Pacman is not working because of the missing certificates. Without propper certificates our Proxy is rejecting any outbound connection. And installing the certificates fails, because I am not root in msys2. As far as I understand the Administrator-Account with the name "Administrator" will be root in msys2 and due our companies User Management - which is similar to the one in many companies - there is no Admistrator named admin account and there never will be and Admistrator named admin account.
Plus I am local admin (with my user name, which is not "Administrator" but something like "abcd"), which in that case seems to make no difference in behavior of msys2.
So first a Fix is needed which enables to install certificates with any local admin or even as ordinary user (hey, normally msys is just used to install a compiler and some toolchain tools. people know what they are doing...)
And Second, as I mentioned any Fix or Solution which enables to install the certificates (install em in msys2 not windows!) can not be done by pacman as pacman needs the solution + installed certificates do work again. So currently we have an hen-egg problem. A Fix needs to be applied, but it can not be distributed via pacman.

And even if I reinstall msys2 from the scratch, I won't be able to install any package with pacman again, as it will also lack our companies certificates. And I won't be able to install the certificates with msys2, as my local admin is not root in terms of msys2...

OK one solution would be just do download a complete msys2 package mirror with windows, and then point pacman to that local repro. but this solution is out of proportion as I don't want to manage and update a local repository by hand.

**Comment by StarWolf3000** [comment]
MSYS2 does not have and does not need something like elevated privileges. There is also no concept like the "root" account. In fact, MSYS2 does not have a user account and permission system.

**Comment by schnedann** [comment]
Well running trust anchor... returns: You are not root, dont't do anything...

![msys2_add_certificate](https://user-images.githubusercontent.com/32240716/68746872-c34fbd80-05f9-11ea-89be-8815f0436e1d.png)

well and yes I even tried to open msys explicit with "run as Administrator" - same output!

Think I found something better last week buts that the best google came up with today: https://sourceforge.net/p/msys2/mailman/message/34142783/

So to summerize:
I am lokal admin - more root I will never become in terms of my companies IT department, I also run msys "as Administrator"
But I can not add certificates!

**Comment by StarWolf3000** [comment]
Invoking `trust anchor --store` calls the script `/usr/libexec/p11-kit/trust-extract-compat`, and if you check it's contents, then you see, this is currently not supported:

```bash
#!/bin/sh

# This script is a placeholder designed to be replaced when this software
# has been customized for distribution. It should be symlinked linked to the
# distribution's update-ca-certificates or update-ca-trust command as
# appropriate. In the future this script will be called when the PKCS#11
# trust module is used to modify trust anchors and related data.

if [ $# -ne 0 ]; then
        echo "usage: trust extract-compat" >&2
        exit 2
fi

uid=$(id -u)
if [ "$uid" != 0 ]; then
        echo "trust: running as non-root user: skip extracting compat bundles" >&2
        exit 0
fi

echo "trust: the placeholder extract-compat command has not been customized by your distribution." >&2

# You can use commands like this to extract data from trust modules
# into appropriate locations for your distribution.
#
# trust extract --format=openssl-bundle --filter=ca-anchors \
#       --overwrite /tmp/openssl-bundle.pem
# trust extract --format=pem-bundle --filter=ca-anchors --overwrite \
#       --purpose server-auth /tmp/server-auth-bundle.pem
# trust extract --format=java-cacerts --filter=ca-anchors --overwrite \
#       --purpose server-auth /tmp/cacerts

exit 1
```

It does nothing at the moment, so importing your certificate with `trust` will not work.

**Comment by schnedann** [comment]
Ok, thats an interesting fact... never thought about debugging the commands itself.
So if that is a dummy, at least there must be some way to install new certificates, right? Any Idea how to accomplish this? Just search for a directory with other certs and copy mine into it sounds too easy...

**Comment by mati865** [comment]
There is `update-ca-trust` command for this purpose.

**Comment by schnedann** [comment]
Thanks for that, at least searching for 

> update-ca-trust

 lead the right way. Soulution found and certifiacates updated... now everything is working again
Still don't understand why the needed parameter is "extract", but hey, it works
